csrc/executor.cpp

```diff
@@ -1606,6 +1607,137 @@ void FusionExecutor::initializeExecutorEntry(
+      // directly copy our buffer into the args buffer. We thus have to
+      // manually downcast each element to fit in the smaller buffer.
+      for (size_t i = 0; i < shape.size(); ++i) {
+        const int32_t shp = static_cast<int32_t>(shape[i]);
+        memcpy(offsets[1] + i * sizeof(int32_t), &shp, sizeof(int32_t));
+      }
+      // In rare cases we have fewer strides than shapes
+      for (size_t i = 0; i < strides.size(); ++i) {
+        const int32_t strd = static_cast<int32_t>(strides[i]);
+        memcpy(offsets[2] + i * sizeof(int32_t), &strd, sizeof(int32_t));
+      }
+    } break;
+    default:
+      NVF_CHECK(0, "Unhandled index type size");
+      break;
+  }
+}
+
+// set the arguments that we'll pass to cuLaunchKernel. This should happen
+// when we change the rank of a tensor or the number of arguments to a kernel.
+// It does not need to happen when only shapes change---use recomputeArgs for
+// that.
+void FusionExecutor::computeArgs(
+    ExecutorEntry& entry,
+    ExpressionEvaluator& expr_eval,
+    const kir::Kernel* kernel) const {
+  FUSER_PERF_SCOPE("Initial GetArgsBuffers");
+
+  const std::vector<Val*>& params = kernel->parameters();
+  entry.args.resize(params.size());
+  entry.arg_ptrs.resize(params.size());
+  const PrimDataType idx_type = kernel->indexType();
+  for (size_t p = 0; p < params.size(); ++p) {
+    entry.args[p] = getKernelArgument(expr_eval, params[p], idx_type);
+    entry.arg_ptrs[p] = entry.args[p].data();
+  }
+}
+
+// Reset the arguments that we'll pass to cuLaunchKernel. This needs to be
+// invoked on every shape change.
+void FusionExecutor::recomputeArgs(
+    ExecutorEntry& entry,
+    ExpressionEvaluator& expr_eval,
+    const kir::Kernel* kernel) const {
+  FUSER_PERF_SCOPE("Recompute GetArgsBuffers");
+  // assert(entry.init && "entry was never initialized");
+
+  const std::vector<Val*>& params = kernel->parameters();
+  const PrimDataType idx_type = kernel->indexType();
+  // assert(entry.args.size() == params.size());
+  // assert(entry.arg_ptrs.size() == params.size());
+  // assert(params.size() >= args.size());
+  for (size_t p = 0; p < params.size(); ++p) {
+    PolymorphicValue pv = expr_eval.evaluate(params[p]);
+    if (pv.is<at::Tensor>() && pv.as<at::Tensor>().is_cuda()) {
+      // GPU tensors are not passed directly: instead we pass a Tensor<type,
+      // rank, rank> struct. The pointer and dimensions are dynamic, but the
+      // types and ranks are actually static (changing the rank or the types
+      // would need to be done via a new FusionDefinition). As such, we created
+      // the Tensor<t, r, r> struct during ::computeArgs, and here we just fill
+      // in the base address, shape, and stride arrays to cover whatever new
+      // tensors we got this round.
+      TensorView* mtv = dynamic_cast<TensorView*>(params[p]);
+      const Val* mdexpr = IrBuilder::metadataExpr(mtv);
+      const PolymorphicValue& tmd = expr_eval.evaluate(mdexpr);
```

Comment:
In the future, we might consider trying to get these 3 lines off the critical path.

In this path, we know we have a tensor and we know we need the 3 fields for that tensor: data, shape, and strides. So ideally we'd just build a:

```c++
struct whatever_t
  void* data;
  IdxType* shape;
  IdxType* strides;
};
```

But we have to go all the way through `PolymorphicValue` and `TensorMetaData` structs, and construct a temporary `Val` here just to get those 3 fields. It would not only be faster to go directly to that struct, it would also be simpler / easier to understand. Maybe a follow-on patch...

Reply:
@zasdfgbnm for viz

Reply:
https://github.com/NVIDIA/Fuser/blob/1e4f744c366f3637c02fda8a197bfe8b38ae77c2/csrc/ir/container.cpp#L318-L325
FYI: tensor metadata is not a temporary Val, it is cached. But still, I agree that skipping these three lines can be faster if the only purpose of getting the tensor metadata is to fill arguments. For other cases, for example when TMA is used, tensor metadata will be evaluated anyway because it is used to create the TMA descriptor.